### PR TITLE
test.py: Refactor retry()

### PR DESCRIPTION
**Pull request by pehala**
* Instead of metamethod that looks at all subclasses, use OOP with super() calls.
* Remove empty `_reset` methods
* Move parameters from reset to `__init__.py`
   * In Python all parameters should ideally be declared in `__init__.py` to get an idea what parameters the object has. Current way of using reset to both reset and initially set parameters was needlessly obfuscated.

`backport/none` because this is an enhancement

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -800,11 +789,21 @@ def __init__(self, test_no: int, shortname: str, suite) -> None:
...
+        self.unidiff: Optional[str] = None
+        self.server_log = None
+        self.server_log_filename = None
+        self.env: Dict[str, str] = dict()
         self._prepare_args(suite.options)
 
-    def _reset(self) -> None:
+    def reset(self) -> None:
         """Reset the test before a retry, if it is retried as flaky"""
+        super().reset()
```
Comment (tchaikov) [inline test.py:806]: nit, probably we can remove the typing annotations in the `reset()` method. now that we already have them in the constructor?

Comment (pehala) [reply test.py:806]: Good catch, removed

**Review by tchaikov — APPROVED** [review]
lgtm

File: test.py [hunk 2]
```diff
@@ -976,7 +970,10 @@ def __init__(self, test_no: int, shortname: str, casename: str, suite) -> None:
         self.xmlout = os.path.join(self.suite.options.tmpdir, self.mode, "xml", self.uname + ".xunit.xml")
         self.server_log: Optional[str] = None
         self.server_log_filename: Optional[pathlib.Path] = None
```
Comment (xtrey) [inline test.py:972]: These strings can be deleted since you are clearing them one more time in next two lines.

**Review by xtrey — APPROVED** [review]

File: test.py (hunk 2 above)
Comment (pehala) [reply test.py:972]: Nice catch :) Fixed

**Review by xtrey — APPROVED** [review]

**Pull request closed without merging**